text/0000-dockerfiles.md

```diff
@@ -0,0 +1,170 @@
+# Meta
+[meta]: #meta
+- Name: Support Dockerfiles
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks.
+
+# Motivation
+[motivation]: #motivation
+
+Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with multi-purpose build-time and runtime Dockerfiles.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### Builder-specified Dockerfiles
+
+A builder image may include any number of "hook" files in `/cnb/hooks.d/`.
+The `pack create-builder` command is used to copy hooks into the builder image via `builder.toml`. E.g.,
+```toml
+[[hooks]]
+name = "app-hook"
+path = "./myhook"
+```
+(Note: this format should not be considered final. Additional fields will be required to mark the format, target, etc. of the hook. These may be defined in subsequent spec PRs / sub-team RFCs.)
+
+Each hook file path must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>(.out|)`, where:
+
+1. `build`, `run`, or empty specify the build-time base image, runtime base image, or both bash images, respectively.
+2. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future.
+3. If the `.out` suffix is present and the file is executable, the hook is executed in the context of the app directory, and its output must match `<format>`.
+4. If the `.out` suffix is not present, the contents of the file must match `<format>`.
+
+Hook files are evaluated (either read or executed) and the resulting Dockerfiles are applied to the build-time and runtime base images in lexographical order by `<name>`.
```

Comment:
Can you please clarify how and when the hooks will be evaluated and/or executed ? As the hook creates images (build and/or run), should we then be executed by another Buildpack step/phase (e.g pre-build, hook, ...) before the step `build` is executed ? Otherwise, it will not be possible to use the new `built` image to by example execute `maven package` as maven is not yet part of the image coming from the hooked dockerfile.

Will it be possible to pass a `build` TOML file containing METADATA to be used by the `hook` to generate the final Dockerfile(s) (e.g. MAVEN version ==> translated by the hook to the command able to install the RPM packages) ?

Reply:
> Can you please clarify how and when the hooks will be evaluated and/or executed ? As the hook creates images (build and/or run), should we then be executed by another Buildpack step/phase (e.g pre-build, hook, ...) before the step build is executed ?

This is stated a few lines below:

> A Dockerfile intended for the runtime base image is applied after the detection phase.
> A Dockerfile intended for the build-time base image is applied before the detection phase.

Detection happens before build. The runtime Dockerfile can be applied at any time before export, because the resulting image isn't used until that phase.

> Will it be possible to pass a build TOML file containing METADATA to be used by the hook to generate the final Dockerfile(s) (e.g. MAVEN version ==> translated by the hook to the command able to install the RPM packages) ?

If the TOML file is in the app directory, then yes, as the hook will execute in the context of the app directory.

Reply:
> A Dockerfile intended for the runtime base image is applied after the detection phase.
> A Dockerfile intended for the build-time base image is applied before the detection phase.

Which inputs parameters will then be passed to the `pre-detect` or ` post-detect` phase ? Why such a question --> To be able to achieve the following scenario: During the execution of the pre or post detect phase, I would like to extract from some METADATA the version of the maven RPM to be installed or JDK requested (java 8, 11, ...) like also if JDK or JRE should be installed. Such METADATA will then be used to set some ENV VAR for the Dockerfile to be executed by kaniko, .... tool.

Reply:
> Which inputs parameters will then be passed to the pre-detect or post-detect phase ?

Executable hooks are run with the app directory set as the working directory, but take no other arguments (aside from env vars set on the base images).

Dockerfiles take the specified ARGS, and inherit env vars from their base images.

> I would like to extract from some METADATA the version of the maven RPM to be installed or JDK requested (java 8, 11, ...) like also if JDK or JRE should be installed.

Where does this metadata come from?

Reply:
@sclevine - I was a little unclear on this point.

>A Dockerfile intended for the runtime base image is applied after the detection phase.
>A Dockerfile intended for the build-time base image is applied before the detection phase.

It says when the Dockerfiles generated will be applied, but I don't see an indication of when the hook will be executed (if it's executable).

For example, it says `A Dockerfile intended for the runtime base image is applied after the detection phase`, but that doesn't necessarily mean that the hook runs after detection. Based on the description, the hook could run prior to detection and the output be held until after.

Is this intentional to allow implementors more room to work? If not, could we clarify the time when the hooks are executed as well?

Reply:
@sclevine - One more question on the execution of hooks.

The docs say that the hook will execute with the app dir set as the working directory and that it does not receive any additional arguments.

For clarification:

1. Would the hook have access to `/cnb/buildpacks/**` in the builder image? If it wanted to, for example, retrieve assets included with buildpacks? 
2. For either `build` or `run`, would there be anything under `/layers/**` yet? Wondering how this would interact with cached layers? Also, because of the timing specified here `The runtime Dockerfile can be applied at any time before export, because the resulting image isn't used until that phase.`, what you could expect? Did the buildpacks run and populate the layers yet? or even if they did would those files be exposed? 

Thanks

Reply:
> could we clarify the time when the hooks are executed as well?

Given:
> Executable hook files must not modify the app directory when executed and may be executed in parallel.
And:
> A Dockerfile intended for the build-time base image is applied before the detection phase.

Build-time hooks could be executed at any point before the Dockerfiles need to be applied (in parallel with buildpack detection would be ideal for performance, but the platform may choose).

It's worth clarifying that runtime hooks must be executed before the build phase (or on a copy of the original source code). I'll add this.

> Is this intentional to allow implementors more room to work? 

I'm trying not to over-specify, given that platforms may want to use their own tooling to apply the hooks, and given that the RFC will be followed by a spec PR. However, I think it's better if the RFC is clear when flexibility is allowed, so I'm happy to add more detail.

> Would the hook have access to /cnb/buildpacks/** in the builder image? If it wanted to, for example, retrieve assets included with buildpacks?

The build-time Dockerfiles run as root on the builder image, so nothing would prevent this. Anything is allowed that doesn't result in a non-compliant builder.

> For either build or run, would there be anything under /layers/** yet? Wondering how this would interact with cached layers?

`/layers` is generally a volume during the build process (so any changes would have no effect) and is generally empty in the runtime base image (so the Dockerfiles probably shouldn't create directories that conflict with buildpack directories).

> Also, because of the timing specified here The runtime Dockerfile can be applied at any time before export, because the resulting image isn't used until that phase., what you could expect? Did the buildpacks run and populate the layers yet? or even if they did would those files be exposed?

Layers are not populated on the runtime base image until the export phase.

Reply:
OK, sounds good. Thanks for clarifying!